test_programs/compile_success_no_bug/encrypted_log_regression/src/main.nr

```diff
@@ -0,0 +1,116 @@
+global MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES: u32 =
+    ENCRYPTED_PAYLOAD_SIZE_IN_BYTES - OVERHEAD_SIZE - PLAINTEXT_LENGTH_SIZE - 1 /* aes padding */;
+
+fn main(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    extended_plaintext: [u8; MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES + PLAINTEXT_LENGTH_SIZE],
+    shift: u32,
+) -> pub [u8; ENCRYPTED_PAYLOAD_SIZE_IN_BYTES] {
+    // unsafe {
+    //     compute_encrypted_log_unconstrained(
+    //         eph_pk_bytes,
+    //         aes_secret,
+    //         incoming_header_ciphertext,
+    //         extended_plaintext,
+    //         shift,
+    //     )
+    // }
+    compute_encrypted_log(
+        eph_pk_bytes,
+        aes_secret,
+        incoming_header_ciphertext,
+        extended_plaintext,
+        shift,
+    )
+}
+
+unconstrained fn compute_encrypted_log_unconstrained<let P: u32, let M: u32>(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    plaintext: [u8; MAX_PRIVATE_LOG_PLAINTEXT_SIZE_IN_BYTES + PLAINTEXT_LENGTH_SIZE],
+    shift: u32,
+) -> [u8; M] {
+    compute_encrypted_log(
+        eph_pk_bytes,
+        aes_secret,
+        incoming_header_ciphertext,
+        plaintext,
+        shift,
+    )
+}
+
+fn compute_encrypted_log<let P: u32, let M: u32>(
+    eph_pk_bytes: [u8; EPH_PK_SIZE],
+    aes_secret: [u8; 32],
+    incoming_header_ciphertext: [u8; HEADER_SIZE],
+    plaintext: [u8; P],
+    shift: u32,
+) -> [u8; M] {
+    let mut encrypted_bytes = [0; M];
+    let mut offset = 0;
+
+    // eph_pk
+    for i in 0..EPH_PK_SIZE {
+        encrypted_bytes[offset + i] = eph_pk_bytes[i];
+    }
+    offset += EPH_PK_SIZE;
+
+    // incoming_header
+    for i in 0..HEADER_SIZE {
+        encrypted_bytes[offset + i] = incoming_header_ciphertext[i];
+    }
+    offset += HEADER_SIZE;
+
+    // Padding.
+    offset += OVERHEAD_PADDING;
+
+    // incoming_body
+    let incoming_body_ciphertext = compute_incoming_body_ciphertext(plaintext, aes_secret);
+    // Then we fill in the rest as the incoming body ciphertext
+    let size = M - offset;
+
+    // NOTE: This made the bytecode size blowup disappear in aztec packages,
+    // but in this reproduction the size seems to be statically known regardless.
+    //let size = M - 32 - HEADER_SIZE - OVERHEAD_PADDING;
+
+    assert_eq(size, incoming_body_ciphertext.len(), "ciphertext length mismatch");
+    for i in 0..size {
+        // NOTE: Adding `shift` makes the index dynamic enough to force ACIR to use memory operations.
+        // Without `shift` it assigns directly to witnesses, which is good, but not what we wanted to reproduce.
+        encrypted_bytes[offset + i + shift] = incoming_body_ciphertext[i];
+    }
+
+    encrypted_bytes
+}
+
+pub fn compute_incoming_body_ciphertext<let P: u32>(
+    plaintext: [u8; P],
+    aes_secret: [u8; 32],
+) -> [u8] {
```

Comment:
This return type is coercing the result to a slice which seems to mess with ACIRgen enough so we get a bunch of unnecessary memory operations. We could either make this return an array or we can have `array_get` instructions with constant indices which act on a `as_slice` instruction "look through it" and act on the underlying array instead.

Reply:
I tried returning an array here, but it didn't make a difference in the bytecode size. Removed this function completely now.